src/pkg/retention/launcher.go

```diff
@@ -14,11 +14,248 @@
 
 package retention
 
-import "github.com/goharbor/harbor/src/pkg/retention/policy"
+import (
+	"errors"
+	"fmt"
+
+	"github.com/goharbor/harbor/src/chartserver"
+	"github.com/goharbor/harbor/src/common/dao"
+	"github.com/goharbor/harbor/src/common/models"
+	"github.com/goharbor/harbor/src/common/utils"
+	"github.com/goharbor/harbor/src/common/utils/log"
+	"github.com/goharbor/harbor/src/core/api"
+	"github.com/goharbor/harbor/src/pkg/retention/policy"
+	"github.com/goharbor/harbor/src/pkg/retention/policy/rule"
+	"github.com/goharbor/harbor/src/pkg/retention/res"
+	"github.com/goharbor/harbor/src/pkg/retention/res/selectors"
+)
+
+var (
+	proMgr  projectManager    = &projectMgr{}
+	repoMgr repositoryManager = &repositoryMgr{}
+)
 
 // Launcher provides function to launch the async jobs to run retentions based on the provided policy.
 type Launcher interface {
 	// Launch async jobs for the retention policy
 	// A separate job will be launched for each repository
-	Launch(policy *policy.Metadata) (string, []string, error)
+	Launch(policy *policy.Metadata) ([]*TaskSubmitResult, error)
+}
+
+// NewLauncher returns an instance of Launcher
+func NewLauncher() Launcher {
+	return &launcher{}
+}
+
+type launcher struct {
+}
+
+func (l *launcher) Launch(ply *policy.Metadata) ([]*TaskSubmitResult, error) {
+	if ply == nil {
+		return nil, errors.New("the policy is nil")
+	}
+	if ply.Scope == nil {
+		return nil, errors.New("the scope of policy is nil")
+	}
+
+	repositoryRules := map[res.Repository]*policy.LiteMeta{}
+	for _, rule := range ply.Rules {
+		repositories, err := getRepositoriesOfRule(ply.Scope, &rule)
+		if err != nil {
+			return nil, err
+		}
+		for _, repository := range repositories {
+			if repositoryRules[*repository] == nil {
+				repositoryRules[*repository] = &policy.LiteMeta{
+					Algorithm: ply.Algorithm,
+				}
+			}
+			repositoryRules[*repository].Rules = append(repositoryRules[*repository].Rules, &rule)
+		}
+	}
+	// TODO init the client
+	var client Client
+	var result []*TaskSubmitResult
+	for repository, rule := range repositoryRules {
+		jobID, err := client.SubmitTask(&repository, rule)
+		result = append(result, &TaskSubmitResult{
+			JobID: jobID,
+			Error: err,
+		})
+		if err != nil {
+			log.Errorf("failed to submit task: %v", err)
+		}
+	}
+	return result, nil
+}
+
+func getRepositoriesOfRule(scope *policy.Scope, rule *rule.Metadata) ([]*res.Repository, error) {
+	var repositoryCandidates []*res.Candidate
+	var err error
+	level := scope.Level
+	switch level {
+	case "system", "project":
+		var projectCandidates []*res.Candidate
+		if level == "system" {
+			// get projects
+			projectCandidates, err = getProjects()
```

Comment:
`getProjects` only need to call one time, but currently called multiple times.

The logic can be simplified, like 

```go
projects :=  make([]*Project,0)

if policy.Scope.Level  == System {
      // only call 1 time
     projects, err = getProjects()
}

for _, r := range policy.Rules {
   subProjects := make([]*res.Candidate, 0)
   switch Scope.Level:{
	 case System:
		proSelector := selectors.Get(r.ScopeSelectors[Project])
		// exchange projects to []*res.Candidates
		subProjects := proSelector.Select(proCandidates)
	case Project:
		subProjects = append(subProjects, policy.Scope.Reference)
   }

   subRepositories := make([]*res.Candidate, 0)
   if Scope.Level = Repository {
	   subRepositories = append(subRepositories, Scope.Reference)
   }else {
	   if len(subProjects) > 0 {
		repoSelector := Selectors.Get(r.ScopeSelectors[Repo])
		for _, proC := range subProjects {
			repoCandidates := getRepos(proC.ID)
			myRepos := repoSelector.Select(repoCandidates)
			subRepositories = append(subRepositories, myRepos...)
		}
	   }
   }
   
   for _, repo := range subRepositories {
	   repoAndRule[repo.Name] = append(repoAndRule[repo.Name], r)
	}
}
```